### ISSUE TITLE
Context is read by components regardless of contextTypes definition

### ISSUE THREAD

**Issue by wojtekmaj** [airbnb/enzyme#1590]
Consider the following, simplified example:

```js
class NotGonnaRead extends Component {
  render() {
    return (
      <p>My favorite color is {this.context.favoriteColor}</p>
    );
  }
}

class WillRead extends Component {
  render() {
    return (
      <p>My favorite color is {this.context.favoriteColor}</p>
    );
  }
}

WillRead.contextTypes = {
  favoriteColor: PropTypes.string,
}

class FavoriteColorProvider extends Component {
  getChildContext() {
    return {
      favoriteColor: 'purple',
    };
  }
}
```

If we render it like so:

```js
<FavoriteColorProvider>
    <NotGonnaRead />
</FavoriteColorProvider>
```

then the output is going to be `My favorite color is `.

Rendering this, however:

If we render it like so:

```js
<FavoriteColorProvider>
    <WillRead />
</FavoriteColorProvider>
```

will correctly output `My favorite color is purple`.

### Current behavior

If we would like to test if NotGonnaRead component is behaving properly (which it isn't), we would need to do something along the lines:

```js
shallow(
  <NotGonnaRead />,
  { context: { favoriteColor: 'purple' } }
);
```

This will render `<p>My favorite color is purple</p>`.

### Expected behavior

This example should render `<p>My favorite color is </p>` as contextTypes are not correctly defined.

#### API

- [x] shallow
- [ ] mount
- [ ] render

#### Version

| library         | version
| ---------------- | -------
| Enzyme           | 3.3.0
| React            |  16.2.0

#### Adapter

- [x] enzyme-adapter-react-16 1.1.1
- [ ] enzyme-adapter-react-15
- [ ] enzyme-adapter-react-15.4
- [ ] enzyme-adapter-react-14
- [ ] enzyme-adapter-react-13
- [ ] enzyme-adapter-react-helper
- [ ] others ( )

**Comment by ljharb** [comment]
This is probably related to #664 - but may also be distinct.

It's possible that enzyme isn't filtering `context` by `contextTypes`, but I'd also expect React to do that for us. Can you check in react 15, and see if the same thing occurs?

**Comment by minznerjosh** [comment]
This was fixed in facebook/react#11922. It was released in `react-test-renderer` v16.3. So I'm going to open a PR shortly that will fix it on enzyme's side for react@<16.3.
